src/prpl.ts

```diff
@@ -12,15 +12,175 @@
+      public entrypoint: string,
+      public pushManifest?: push.PushManifest) {
   }
+
+  /**
+   * Order builds with more capabililties first -- a heuristic that assumes
+   * builds with more features are better. Ties are broken by the order the
+   * build appeared in the original configuration file.
+   */
+  compare(that: Build): number {
+    if (this.requirements.size !== that.requirements.size) {
+      return that.requirements.size - this.requirements.size;
+    }
+    return this.configOrder - that.configOrder;
+  }
+
+  /**
+   * Return whether all requirements of this build are met by the given client
+   * browser capabilities.
+   */
+  canServe(client: Set<capabilities.BrowserCapability>): boolean {
+    for (const r of this.requirements) {
+      if (!client.has(r)) {
+        return false;
+      }
+    }
+    return true;
+  }
+}
+
+function loadBuilds(config: ProjectConfig|undefined, root: string): Build[] {
+  const builds: Build[] = [];
+  const entrypoint = (config ? config.entrypoint : null) || 'index.html';
+
+  if (!config || !config.builds) {
+    // No builds were specified. Try to serve an entrypoint from the root
+    // directory, with no capability requirements.
+    console.warn(`warning: no builds configured`);
+    builds.push(new Build(0, new Set(), path.join(root, entrypoint)));
+
+  } else {
+    for (let i = 0; i < config.builds.length; i++) {
+      const build = config.builds[i];
+      if (!build.name) {
+        console.warn(`warning: build at offset ${i} has no name; skipping`);
+        continue;
+      }
+
+      // TODO Push manifest location should be configurable.
+      const pushManifestPath =
+          path.join(root, build.name, 'push-manifest.json');
+      let pushManifest;
+      if (fs.existsSync(pushManifestPath)) {
+        console.info(`detected push manifest "${pushManifestPath}"`);
+        const pushManifestData =
+            JSON.parse(fs.readFileSync(pushManifestPath, 'utf8')) as
+            push.PushManifestData;
+        pushManifest = new push.PushManifest(pushManifestData);
+      }
+
+      builds.push(new Build(
+          i,
+          new Set(build.browserCapabilities),
+          path.join(root, build.name, entrypoint),
+          pushManifest));
+    }
+  }
+
+  // Sort builds by preference in case multiple builds could be served to
+  // the same client.
+  builds.sort((a, b) => a.compare(b));
+
+  // Sanity check.
+  let hasFallback = false;
+  for (const build of builds) {
+    hasFallback = hasFallback || build.requirements.size === 0;
+    const requirements = Array.from(build.requirements.values());
+    console.info(
+        `registered entrypoint "${build.entrypoint}" with capabilities ` +
+        `[${requirements.join(',')}]`);
+    if (!fs.existsSync(build.entrypoint)) {
+      console.warn(`warning: entrypoint "${build.entrypoint}" does not exist`);
+    }
+  }
+  if (!hasFallback) {
+    console.warn(`warning: no hasFallback build registered`);
```

Comment:
This warning isn't super clear. I'd mention explicitly that there's no build without requirements, so the site will not work in all browsers.